README.md
# Jump Rope Database
Create a database that is system agnostic and able to support multiple organizations


CHANGELOG
# Jump Rope Database Release Notes

## 0.0.1 (15 Mar 2023)

* Initial release


tests/test_app.py
from jumpropedatabase.databasing import database
import os


dir_data = r"/Users/"+os.getlogin()+r"/Library/Application Support/"
def test_creating_database():
    "An initial test for the app"
    
    assert 1 + 1 == 2


tests/test_database_unit.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Mar 29 18:45:54 2023

@author: haydenthomas
"""

import os
from jumpropedatabase.databasing import database
from sqlite3 import Connection
from pandas import read_sql_query


# main variables
path = os.path.curdir
file = "tutorial.db"
table_name = "test"
data_input = [
    ("Monty Python Live at the Hollywood Bowl", 1982, 7.9),
    ("Monty Python's The Meaning of Life", 1983, 7.5),
    ("Monty Python's Life of Brian", 1979, 8.0),
]
query = "SELECT * FROM test WHERE year > 1900 ORDER BY year DESC"


# remove sql database if it exists
if os.path.isfile(path+"/"+file):
    os.remove(path+"/"+file)

def test_creating_database():
    '''test_creating_database tests creating a database called tutorial.db
    

    Returns
    -------
    None.

    '''
    
    con = database.conn_database(path, file)
    assert type(con) == Connection


def test_make_table():
    '''test_make_table creates a table in the database called "test" with the 
    columns title, year, and score
    

    Returns
    -------
    None.

    '''
    con = database.conn_database(path, file)
    database.make_table(con,table_name,["title","year","score"])
    data = read_sql_query("SELECT * FROM "+table_name, con)
    assert data.empty == True


def test_add_data():
    '''test_add_data adds data to the database called test
    

    Returns
    -------
    None.

    '''
    
    con = database.conn_database(path, file)
    database.add_data_to_table(con,table_name,data_input)
    data = read_sql_query("SELECT * FROM "+table_name, con)
    assert data.empty == False


def test_read_data():
    '''test_read_data tests reading in the data and verifies that the first
    entry of the year is the maximum date as defined by the query.
    

    Returns
    -------
    None.

    '''
    con = database.conn_database(path, file)
    data_out = database.read_data_from_table(con, query)
    print(data_out.year[0])
    assert data_out.year[0] == data_out.year.max()


    
    

src/jumpropedatabase/utilities/get_data_path.py
# -*- coding: utf-8 -*-
"""
Created on Mon Mar 20 17:32:29 2023

@author: thomahm1
"""

from os import path
from platform import system

def get_data_path():
    '''
    get_data_path identifies the data path depending on if the system is
    Windows or IOS.

    Returns
    -------
    valid_system : bool
        boolean value to determine if it is an acceptable system.
    system_data_path : str
        path to install and store data.

    '''
                  
    valid_system = False
    system_data_path = str()
    if system().find("Windows") > -1:
        valid_system = True 
        system_data_path = 'C:\\Program Files (x86)\\jump rope database\\'
    elif system().find("Darwin") > -1:
        valid_system = True
        system_data_path = '~/Library/Application Support/jump rope database'
    else:
        # immediately return False and empty string
        return valid_system, system_data_path
    
    
    return valid_system, system_data_path

src/jumpropedatabase/utilities/create_contents.py
# -*- coding: utf-8 -*-
"""
Created on Sat Mar 25 07:41:52 2023

@author: thomahm1
"""

from os import path, mkdir, rmdir, remove
import json

def make_dir(path_name:str):
    '''Creates a directory assuming it does not already exist.
    

    Parameters
    ----------
    path_name : str
        Path to create the folder for storing data. For example database 
        folders, executable folder, etc.

    Returns
    -------
    None.

    '''
    if not path.exists(path_name):
        mkdir(path_name)
        print("Created " + path_name)
    else:
        print(path_name + ' already exists')
    
def make_json_file(path_name:str, json_name:str):
    '''Creates a JSON file in the user defined path
    

    Parameters
    ----------
    path_name : str
        Path to create the JSON file.
    json_name : str
        Name of the JSON file.

    Returns
    -------
    None.

    '''
    
    file_name = path.join(path_name, json_name)
    if not path.isfile(file_name):
        open(file_name, "w+")
        print("Created " + file_name)
    else:
        print(file_name + ' already exists')

def add_json_data(path_name:str, json_name:str, new_json_data:dict):
    
    file_name = path.join(path_name, json_name)
    json_file_data = json.loads(file_name)
    
    

if __name__ == "__main__":
    # test creating a folder
    folder_test = "./test"
    make_dir(folder_test)
    
    # test an already created folder
    make_dir(folder_test) # should only generate a 
    
    # test creating a JSON file
    file_name = "test.json"
    make_json_file(folder_test, file_name)
    
    # test trying to create an existing JSON file
    make_json_file(folder_test, file_name)
    
    remove(path.join(folder_test, file_name))
    rmdir(folder_test)

src/jumpropedatabase/databasing/database.py
from sqlite3 import connect, Connection
import os
from pandas import read_sql_query


def conn_database(path:str, name:str):
    '''
    conn_database creates a database in the user defined path and name

    Parameters
    ----------
    path : str
        path to create the database
    name : str
        name of the database file

    Returns
    -------
    con : sqlite3.Connection
        connection to the user specified database

    '''
    if not name.endswith(".db"):
        name = name+".db"
    con = connect(f"{name}")
    
    return con

def make_table(con:Connection,table_name:str,column_names:list):
    '''
    make_table creates a table in the user defined database with the table 
    name and column names defined by the user

    Parameters
    ----------
    con : Connection
        Connection to the database (assumed to be open).
    table_name : str
        Name for the table.
    column_names : list
        List of strings for the name of each column.

    Returns
    -------
    None.

    '''
    cur = con.cursor()
    
    column_name_string = ""
    for i in range(0,len(column_names)):
        if i > 0:
            column_name_string = column_name_string+","+column_names[i] 
        else:
            column_name_string=column_names[i]
    
    print(f"CREATE TABLE {table_name}({column_name_string})")
    cur.execute(f"CREATE TABLE {table_name}({column_name_string})")

def add_data_to_table(con:Connection,table_name:str,data_in:list):
    '''
    

    Parameters
    ----------
    con : Connection
        Connection to database.
    table_name : str
        Name of the table to get data from.
    data_in : list
        List of tuples of the length of the number of columns in the dataset.
        Adds the number of rows as there are columns in the dataset.

    Returns
    -------
    None.

    '''
    cur = con.cursor()
    
    # create string for each column to show what data to insert
    column_string = ""
    for i in range(0,len(data_in[0])):
        if i > 0:
            column_string = column_string+",?"
        else:
            column_string="?"
            
    # add data to table
    cur.executemany(f"""INSERT INTO {table_name} 
                    VALUES ({column_string})""",data_in)
    
    con.commit()

def read_data_from_table(con:Connection, query:str):
    '''
    

    Parameters
    ----------
    con : Connection
        Connection to database.
    query : str
        Query of the database with the form SELECT ? FROM ? WHERE ? GROUP BY ?
        .

    Returns
    -------
    data_out : TYPE
        DESCRIPTION.

    '''
    
    
    data_out = read_sql_query(query, con)
    return data_out
    



if __name__ == "__main__":
    path = os.path.curdir
    file = "tutorial.db"
    
    if os.path.isfile(path+"/"+file):
        os.remove(path+"/"+file)
    con = conn_database(path, file)
    make_table(con, "test",["title","year","score"])
    
    data = [
        ("Monty Python Live at the Hollywood Bowl", 1982, 7.9),
        ("Monty Python's The Meaning of Life", 1983, 7.5),
        ("Monty Python's Life of Brian", 1979, 8.0),
    ]
    add_data_to_table(con,"test",data)
    
    query = "SELECT * FROM test WHERE year > 1900 ORDER BY year DESC"
    data_out = read_data_from_table(con, query)
    
    con.close()
